todo_app/tests/test_client.py

```diff
@@ -0,0 +1,82 @@
+from dotenv import load_dotenv, find_dotenv
+from todo_app import app
+from todo_app.data.trello_helper import TrelloHelper
+import os, requests, pytest
+
+@pytest.fixture
+def client():
+    file_path = find_dotenv('.env.test')
+    load_dotenv(file_path, override=True)
+
+    test_app = app.create_app()
+
+    with test_app.test_client() as client:
+        yield client
+
+
+
+class StubResponse():
+    def __init__(self, fake_response_data):
+        self.fake_response_data = fake_response_data
+
+    def json(self):
+        return self.fake_response_data
+    
+
+class DataStubResponse():
+    def __init__(self, data_from_call):
+        self.data_from_call = data_from_call
+
+    def json(self):
+        return self.data_from_call
+
+def get_stub(url, params={}):
+    test_board_id = os.environ.get('TRELLO_BOARD_ID')
+    
+    if url == f'https://api.trello.com/1/boards/{test_board_id}/lists':
+        fake_response_data = [{
+            'id': 'to_do_list_id',
+            'name': 'To Do',
+            'cards': [{'id':'to_do_card_id','name': 'To_do Test card','shortUrl': 'http://myurl_to_do'}]
+        },{
+            'id': 'doing_list_id',
+            'name': 'Doing',
+            'cards': [{'id':'doing_card_id','name': 'Doing Test card','shortUrl': 'http://myurl_doing'}]
+        },{
+            'id': 'done_list_id',
+            'name': 'Done',
+            'cards': [{'id':'done_card_id','name': 'Done Test card','shortUrl': 'http://myurl_done'}]
+        }]
+
+        return StubResponse(fake_response_data)
+
+    raise Exception(f'Integration test did not expect URL "{url}"')
+
+def put_stub(url, data, headers):
+    test_card_id_for_moving = 'to_do_card_id'
+    move_to_list = TrelloHelper._LIST_DONE
+    if not url == f'https://api.trello.com/1/cards/{test_card_id_for_moving}':
+        raise Exception(f'Integration test did not expect URL "{url}"')
+
+def test_index_page(monkeypatch, client):
+    monkeypatch.setattr(requests, 'get', get_stub)
+    response = client.get('/')
+
+    decoded_data = response.data.decode()
+
+    assert response.status_code == 200
+    assert 'To_do Test card' in decoded_data
+    assert 'Doing Test card' in decoded_data
+    assert 'Done Test card' in decoded_data
+
+def test_complete_item(monkeypatch, client):
+    # CALLING GET HERE AS NEED TO INITIALISE LIST ID VARIABLES IN TRELLO HELPER
+    monkeypatch.setattr(requests, 'get', get_stub)
+    monkeypatch.setattr(requests, 'put', put_stub)
+    get_response = client.get('/')
+    post_response = client.post('/completeItem/to_do_card_id')
+    # NOTE - I WANTED TO TEST THE CONTENT OF THE DATA IN THE POST HERE BUT COULDNT FIGURE OUT HOW TO DO IT
+    assert post_response.status_code == 302
```

Comment:
This depends a bit on what you're actually trying to capture!

## Testing content sent to the stub:

If you want to test the content that was sent to the _stub_, i.e. to make sure it actually received the right message, then we could try to capture that within the stub. One (incomplete) approach might be:

Set up our stub to record every call that it gets triggered with - e.g. using a decorator:
```python
put_calls = []

def record(func):
    @functools.wraps(func)
    def wrapped(*args, **kwargs):
        put_calls.append({"args": args, "kwargs": kwargs})
        return func(*args, **kwargs)
    return wrapped

@record
def put_stub(url, data, headers):
```

And we can then make an assertion about those here:
```python
    assert len(put_calls) == 1
    # more useful assertions likely exist!
    assert 'to_do_card_id' in put_calls[0]['args'][0]
```

Only caveat would be that we likely want to implement a clean-up strategy else our tests will quickly start interfering with each other:
```
@pytest.fixture(autouse=True)
def setup():
    global put_calls
    put_calls = []
```

## Mocking the item showing on the homepage

If on the other hand you're thinking about testing that the item appears on the homepage correctly after the change, we'll need to tweak our stubs so that they can see when that changes. Here we're going to have to slightly restructure so that we have an idea about a data store that can change, so potentially like:

```python

cards = {
    'to_do_card_id': {'id':'to_do_card_id','name': 'To_do Test card','shortUrl': 'http://myurl_to_do', 'idList': 'to_do_list_id'},
    'doing_card_id': {'id':'doing_card_id','name': 'Doing Test card','shortUrl': 'http://myurl_doing', 'idList': 'doing_list_id'},
    'done_card_id': {'id':'done_card_id','name': 'Done Test card','shortUrl': 'http://myurl_done', 'idList': 'done_list_id'}
}

def get_stub(url, params={}):
    test_board_id = os.environ.get('TRELLO_BOARD_ID')
    
    if url == f'https://api.trello.com/1/boards/{test_board_id}/lists':
        fake_response_data = [{
            'id': 'to_do_list_id',
            'name': 'To Do',
            'cards': [card for card in cards.values() if card['idList'] == 'to_do_list_id']
        },{
            'id': 'doing_list_id',
            'name': 'Doing',
            'cards': [card for card in cards.values() if card['idList'] == 'doing_list_id']
        },{
            'id': 'done_list_id',
            'name': 'Done',
            'cards': [card for card in cards.values() if card['idList'] == 'done_list_id']
        }]

        return StubResponse(fake_response_data)
```

And our put stub could then:
```python
def put_stub(url, data, headers):
    test_card_id_for_moving = 'to_do_card_id'
    move_to_list = TrelloHelper._LIST_DONE
    if not url == f'https://api.trello.com/1/cards/{test_card_id_for_moving}':
        raise Exception(f'Integration test did not expect URL "{url}"')
    cards[test_card_id_for_moving]['idList'] = 'done_list_id'
```

Which we could then test something like:
```python
    post_response = client.post('/completeItem/to_do_card_id', follow_redirects = True)
    assert post_response.status_code == 200
    # There may be a better way to test this is in the right column - possibly by adding some test-tagging in HTML
    # but this will suffice here hopefully!
    assert '<input type="submit" name="Reopen" value="Reopen"> To_do Test card' in post_response.data.decode()
```